### PR TITLE
feat(repo-graph): add workspace dependency graph with hook wiring and hardening

### LINKED ISSUE #470: Repo Map / Code Graph: hardening + unwired follow-ups from #469 review
Tracking issue for follow-up work surfaced by the parallel sub-agent review of #469 (Repo Map / Code Graph). None are blocking — all confirmed P0/P1/P2 findings from that review are fixed on `main` via #469. The items below are quality-of-life hardening, unwired functionality from the original plan, and small test-coverage gaps.

### 1. Auto-build hook on session start
The original plan included a `src/hooks/repo-graph-builder.ts` that would build/refresh `.swarm/repo-graph.json` automatically on session start and incrementally update it after coder write-tool calls. We shipped without this — agents must currently invoke `repo_map action="build"` manually (typically the architect during planning), and incremental updates only happen if a caller explicitly invokes `updateGraphIncremental`.

**Acceptance:**
- New hook attached to session start that builds the graph if missing or older than the freshness window.
- Hook attached to `tool.execute.after` for `WRITE_TOOL_NAMES` that calls `updateGraphIncremental` for the touched paths.
- In-memory graph cached at module scope; disk read only on cold start.

### 2. Tree-sitter migration (deferred for performance)
The current import + symbol extractors are regex-based. The plan called for Tree-sitter via the existing `loadGrammar()` runtime. Regex was chosen because the regex implementation completes a 999-file/5089-edge build in ~2 s, but rare edge cases (dynamic imports, unusual quoting, member-expression false positives — see #469's F1.1) keep surfacing. Worth re-evaluating once the graph layer has more callers.

### 3. Workspace-internal symlink escape (F3 follow-up)
`updateGraphIncremental` validates that the relative path doesn't escape the workspace, but a symlink *inside* the workspace pointing outside (e.g. `src/link.ts → /tmp/outside/evil.ts`) currently passes validation and `processFile` follows the symlink. Not exploitable in normal use, but a defense-in-depth `fs.realpath` check inside the file walker / `processFile` would close the gap.

### 4. `.swarm/repo-graph.json` symlink defense (F5 follow-up)
`saveGraph` `lstat`s the `.swarm/` directory. The graph file inside `.swarm/` is not similarly checked — if the file is a pre-existing symlink, `renameSync` replaces it in-place on Linux (benign), but defense-in-depth `lstatSync` on the destination file before rename would lock this down across platforms.

### 5. F11 test strength
The current F11 test asserts UUID-shape on orphan tmpfiles produced during a forced-rename failure. In the happy path no orphan exists, so the assertion is vacuous. Intercept `fs.writeFileSync` (or use `fs.promises` mocking) to capture the actual tmp path and assert the UUID pattern unconditionally.

### 6. `buildRepoGraph` truncation telemetry (F4 follow-up)
When `findSourceFiles` returns more than `DEFAULT_MAX_FILES`, files are silently sliced. Add a debug-level log noting `total → cap` so users on huge monorepos know files were dropped. Also clarify whether `maxFiles: 0` means "no files" or "use default" — currently it means "use default" because of the `> 0` guard.

### 7. `normalizeGraphPath` edge cases (F8 follow-up)
- Mid-path `/./` segments (e.g. `util/./other.ts`) are not collapsed; could cause graph-key mismatches if a caller produces such paths.
- Windows paths starting with `\.\` become `/.//util.ts` after backslash conversion and the leading `/` survives. Edge case, but worth either documenting the input contract or normalizing more aggressively (`path.posix.normalize`).

### 8. `getBlastRadius` long-chain regression test (F9 follow-up)
The existing tests cover `maxDepth=1` and `maxDepth=2` with a 2-hop chain. Add a test with a chain longer than `maxDepth` to lock in the bound (`depthReached <= maxDepth` even when there are more transitive importers).

### 9. `path.join` Windows-targeted unit test (F7 follow-up)
We use `path.join` for index-file lookups so `lib/index.ts` becomes `lib\index.ts` on Windows. Add a Windows-only unit test (or assertion behind a `process.platform === 'win32'` skip) that verifies no `/` appears between the directory and `index.ts` in resolved paths.

### 10. Additional regression tests for known-bypass shapes
- F1: multi-line `require(\n "./x"\n)` — works today but is unverified.
- F1: `from"./x"` (no whitespace before quote) — currently silently dropped.
- F2: side-effect import line numbers — `import './a';\nimport './b';` should report lines 1 and 2.

### 11. Pre-existing `tests/unit/hooks` mock-conflict failures
Running `bun --smol test tests/unit/hooks --timeout 60000` reports 1633 failures both on `main` and on every feature branch — mock state leaks between files. Per-file loops (`for f in ...; do bun --smol test "$f"; done`) work fine. Worth either documenting in CONTRIBUTING.md or fixing the mock teardown so the batch run stops being noise.

---

These items were all surfaced by the post-implementation parallel-verifier sub-agents from #469's review. They are tracked here so the actionable diffs from that PR don't grow into a sprawling follow-up commit.

### PR DESCRIPTION
**Pull request by zaxbysauce**
## Summary
- Add workspace dependency graph system with node/edge model, safe load/save with atomic writes, symlink boundary validation, and cross-platform path normalization (closes #470)
- Wire startup init() hook and write-triggered incremental update hook into tool handler chain via safeHook() in src/index.ts
- Expand import parsing to cover re-export patterns and dynamic import() expressions; harden with workspace boundary validation and mock.module elimination via dependency injection

## Test plan
- [x] Tier 1: un run typecheck and unx biome ci . pass (0 errors, 102 pre-existing warnings)
- [x] Tier 2: All 10 repo-graph tool test files pass per-file isolation (0 failures); hooks/cli/commands/config batch passes (3683 pass; 2245 pre-existing failures unrelated to this change)
- [x] Tier 4: Security tests 97/97 pass; adversarial tests 652/674 pass (22 pre-existing failures unrelated to this change)
- [x] Tier 5: un run build succeeds
- [x] All 148 repo-graph tests pass together in a single process (0 failures, 1 intentional Windows skip)
- [x] Review council (security, architecture, test quality, coder self-assessment) completed — findings tracked in #474

## Pre-existing test failures
The batch test run shows 2245 failures in hooks/cli/commands/config and 22 adversarial failures. These are all pre-existing and unrelated to the repo-graph changes. All repo-graph-specific tests pass in isolation and together.

### CODE REVIEW

File: src/tools/repo-graph.ts [hunk 1]
```diff
@@ -0,0 +1,1459 @@
...
+	}
+	if (containsControlChars(workspace)) {
+		throw new Error('Invalid workspace: control characters detected');
+	}
+	if (containsPathTraversal(workspace)) {
+		throw new Error('Invalid workspace: path traversal detected');
+	}
+	if (workspace.startsWith('/') || workspace.startsWith('\\')) {
+		throw new Error('Invalid workspace: absolute path not allowed');
+	}
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply src/tools/repo-graph.ts]: Verified against the code. This is a real bug: \alidateWorkspace\ (line 131-136) rejects absolute paths, but the hook is wired at \src/index.ts:164\ with \ctx.directory\ (absolute). The \init()\ try/catch swallows the error silently. The fix is to have the hook convert the absolute path to relative before passing to graph functions, since the downstream API (\getGraphPath\, \saveGraph\, \loadGraph\) all construct \.swarm/\ paths from relative roots. Will fix.

File: src/tools/repo-graph.ts [hunk 2]
```diff
@@ -0,0 +1,1459 @@
...
+			const normalizedRoot = path.normalize(realRoot);
+
+			// Ensure result is within workspace using real path boundaries
+			if (
+				!normalizedResolved.startsWith(normalizedRoot + path.sep) &&
+				normalizedResolved !== normalizedRoot
+			) {
+				return null;
+			}
+			return resolved;
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply src/tools/repo-graph.ts:352]: Verified against the code. This is a real data quality bug: \
esolveModuleSpecifier\ (line 287) does \path.resolve(sourceDir, specifier)\ which produces extensionless paths like \/path/to/utils\. Node keys in the graph use paths WITH extensions from \indSourceFiles\. So edges have dangling targets. The claim about \updateGraphForFiles\ causing rebuild loops is exaggerated — it's just incomplete edges, not a functional failure. Fix: try common extensions (\.ts\, \.tsx\, \.js\, \.jsx\, \.mjs\, \.cjs\, \.py\, \.json\) when the resolved path doesn't exist. Will fix.
